perf_bench_test.go

```diff
@@ -24,127 +24,346 @@ package main
+		require.NoError(t, err, "Function returned error")
+		require.Equal(t, resp.Payload, "Hello, record_response!")
 	}
+}
 
-	// Boot VMs
-	for i := 0; i < *vmNum; i++ {
+// Boot a range of VMs with given images
+func bootVMs(t *testing.T, images []string, startVMID, endVMID int) {
+	for i := startVMID; i < endVMID; i++ {
 		vmIDString := strconv.Itoa(i)
-		_, err := funcPool.AddInstance(vmIDString, images[funcs[i%len(funcs)]])
+		_, err := funcPool.AddInstance(vmIDString, images[i%len(images)])
 		require.NoError(t, err, "Function returned error")
 	}
+}
 
-	if !*isWithCache && *isColdStart {
-		log.Info("Profile cold start")
-		dropPageCache()
-	}
+// Inject many requests per second to VMs and profile
+func loadAndProfile(t *testing.T, images []string, vmNum, targetRPS int, isSyncOffload bool) map[string]float64 {
+	var (
+		vmID, requestID             int
+		invokExecTime, realRequests int64
+		vmGroup                     sync.WaitGroup
+		isProfile                   = false
+		timeInterval                = time.Duration(time.Second.Nanoseconds() / int64(targetRPS))
+		injectDuration              = *warmUpTime + *profileTime + *coldDownTime
+		totalRequests               = injectDuration * float64(targetRPS)
+		remainingRequests           = totalRequests
+		ticker                      = time.NewTicker(timeInterval)
+		profiler                    = profile.NewProfiler(injectDuration, *profilerInterval, vmNum, *profilerLevel, *profilerNodes, "profile")
+	)
 
-	ticker := time.NewTicker(timeInterval)
-	var vmGroup sync.WaitGroup
-	done := make(chan bool, 1)
+	const (
+		averageExecutionTime  = "average-execution-time"
+		realRequestsPerSecond = "real-requests-per-second"
+	)
+
+	tStart := time.Now()
+	err := profiler.Run()
+	require.NoError(t, err, "Run profiler returned error")
+	go profileControl(vmNum, &isProfile, profiler)
 
-	for totalRequests > 0 {
-		select {
-		case <-ticker.C:
-			totalRequests--
+	for remainingRequests > 0 {
+		if tickerT := <-ticker.C; !tickerT.IsZero() {
 			vmGroup.Add(1)
+			remainingRequests--
 
-			funcName := funcs[vmID%len(funcs)]
-			imageName := images[funcName]
-			vmIDString := strconv.Itoa(vmID)
+			imageName := images[vmID%len(images)]
 
-			tStart := time.Now()
-			go serveVM(t, tStart, vmIDString, imageName, &vmGroup, isSyncOffload)
+			go serveVM(t, &vmGroup, vmID, requestID, imageName, isSyncOffload, &isProfile, &invokExecTime, &realRequests)
+			requestID++
 
-			vmID = (vmID + 1) % *vmNum
-		case <-done:
-			ticker.Stop()
+			vmID = (vmID + 1) % vmNum
 		}
 	}
-
-	done <- true
+	ticker.Stop()
 	vmGroup.Wait()
+	profiler.SetTearDownTime()
+	log.Debugf("All VM returned in %d Milliseconds", time.Since(tStart).Milliseconds())
+
+	// Collect results
+	serveMetrics := make(map[string]float64)
+	serveMetrics[averageExecutionTime] = float64(invokExecTime) / float64(realRequests)
+	serveMetrics[realRequestsPerSecond] = float64(realRequests) / (profiler.GetTearDownTime() - profiler.GetWarmupTime())
```

Comment:
BUG
`GetWarmupTime()` does not correspond to the real warmup time, hence the real RPS would be computed incorrectly

Reply:
Done